internal/controllers/machineset/machineset_preflight.go

```diff
@@ -0,0 +1,220 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package machineset
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/blang/semver"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/pointer"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1beta1"
+	"sigs.k8s.io/cluster-api/controllers/external"
+	"sigs.k8s.io/cluster-api/feature"
+	"sigs.k8s.io/cluster-api/internal/contract"
+)
+
+type preflightCheckError *string
+
+// preflightFailedRequeueAfter is used to requeue the MachineSet to re-verify the preflight checks if
+// the preflight checks fail.
+const preflightFailedRequeueAfter = 15 * time.Second
+
+func (r *Reconciler) runPreflightChecks(ctx context.Context, cluster *clusterv1.Cluster, ms *clusterv1.MachineSet, action string) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	// If the MachineSetPreflightChecks feature gate is disabled return early.
+	if !feature.Gates.Enabled(feature.MachineSetPreflightChecks) {
+		return ctrl.Result{}, nil
+	}
+
+	skipped := skippedPreflightChecks(ms)
+	// If all the preflight checks are skipped then return early.
+	if skipped.Has(clusterv1.MachineSetPreflightCheckAll) {
+		return ctrl.Result{}, nil
+	}
+
+	// If the cluster does not have a control plane reference then there is nothing to do. Return early.
+	if cluster.Spec.ControlPlaneRef == nil {
+		return ctrl.Result{}, nil
+	}
+
+	// Get the control plane object.
+	controlPlane, err := external.Get(ctx, r.Client, cluster.Spec.ControlPlaneRef, cluster.Namespace)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform %q: failed to perform preflight checks: failed to get ControlPlane %s", action, klog.KRef(cluster.Spec.ControlPlaneRef.Namespace, cluster.Spec.ControlPlaneRef.Name))
```

Comment:
q: should this be treated as error or simply became a no-op given that preflight checks do not apply if there is no CP?

Reply:
Same here. This is the case where we have a control plane ref in Cluster but the control plane does not exist. I would prefer surfacing the error over just ignoring it.